doc/developer/design/20240609_error_handling.md

```diff
@@ -0,0 +1,413 @@
+ `offset`       | `bigint`       | No       | The offset of the Kafka message that failed to decode.
+ `timestamp`    | `timestamp`    | No       | The timestamp of the Kafka message that failed to decode.
+ `key`          | `bytea`        | Yes      | The key bytes of the Kafka message that failed to decode.
+ `value`        | `bytea`        | Yes      | The value bytes of the Kafka message that failed to decode.
+
+The DLQ table works like a subsource. It can be queried via `SELECT`, subscribed
+to via `SUBSCRIBE`, and used as the upstream relation for a Kafka sink.
+
+The DLQ table is append only. Allowing users to specify a retention policy on
+the DLQ table is left to future work. Users cannot issue `INSERT`, `UPDATE`, or
+`DELETE` commands against the DLQ.
+
+When a DLQ table is attached to a source, no errors are emitted to the source's
+*errs stream* (see the [error handling section][error-handling-docs] of the
+`compute::render` module).
+
+**Open questions:**
+
+  * Is there a better name for the option than `REDIRECT ERRORS`?
+  * What type gets reported in the system catalog for DLQ tables? Is it really a
+    table? Or is it a source or a subsource?
+  * Is it useful to report the `key` and the `value` in the DLQ relation? Most
+    of the time they will be unreadable bytes.
+  * Can the `REDIRECT ERRORS` option be added to an existing source (i.e., via
+    `ALTER SOURCE`)? If so, what is the behavior? Does it pick up for all new
+    messages?
+  * Does the `mz_timestamp` column make sense? Is there something better
+    to call this column? Do the semantics make sense? (Including an
+    `ingested_at` wall clock timestamp would be nice, but that wouldn't
+    be a deterministic function of reclocking.)
+
+#### Upsert errors
+
+While the DLQ is useful for alerting users to data quality issues, it doesn't
+help users answer the question "does my upsert source currently have any keys
+whose most recent value failed to decode?"
+
+To solve this problem, we propose the addition of an `INLINE ERRORS` option
+to `ENVELOPE UPSERT`:
+
+```sql
+CREATE SOURCE src TO KAFKA CONNECTION kconn ...
+ENVELOPE UPSERT (
+    VALUE DECODING ERRORS = ({INLINE | PROPAGATE}, ...)
+)
+```
+
+The default behavior, which matches today's behavior, is `PROPAGATE`. Value
+decoding errors are propagated to the DLQ table or the source's *err stream*,
+whichever is active.
+
+When the `INLINE` behavior is specified, the source's relation gains a nullable
+column named `error` with a type of `record(description: text, code: text)`. If
+the most recent value for a key has been successfully decoded, this column will
+be `NULL`. If the most recent value for a key was not succesfully decoded, this
+column will contain details about the error. Additionally, the value is forced
+into a single nullable column named `value` with a type reflecting the
+configured value format—flattening of record-type values into top-level columns
+does not occur.
```

Comment:
> Additionally, the value is forced into a single nullable column named `value`

Why is this change necessary? Couldn't we just make all the columns nullable and put null in all value columns when there is an error? Or better yet, if the error happened only when decoding the value of one of the columns, put null only in that column. (Note that distinguishing real nulls from error nulls would still be possible, because one could additionally look at the error column.)

Note that unpacking a record is currently not a very efficient operation, because if you write
```
SELECT rec.a, rec.b, rec.c, rec.d, rec.e
```
we compile this to 5 `record_get` function calls, and each `record_get` call gets the full record, and [it iterates through the fields, decoding each `Datum`, to get to the requested one](https://github.com/MaterializeInc/materialize/blob/9a1425086b77081f1878af192010d7aa3906a5d6/src/expr/src/scalar/func/impls/record.rs#L155).

Reply:
Related question -- is this `value` column also at the top-level of the source relation? So we are introducing two new columns to each upsert source, `error` and `value`?

Reply:
Yeah, exactly. So if the upsert source used to contain `offset, timestamp, a, b`, it would now contain `offset, timestamp, value, error`, and `value` would be a record type containing columns `a` and `b`.

Reply:
Started a thread on potential performance improvement of record unpacking: https://materializeinc.slack.com/archives/C0761MZ3QD9/p1718655594016239